File: toki/solutions/toki_pair.py
import sys

sys.setrecursionlimit(100000)

CONSONANTS, VOWELS, ILLEGALS = 'mnptkswjl', 'aeiou', 'wu wo ji ti nn nm'

def solve(W: str) -> str:
    """
    Considers if W is a valid syllable or if W can be broken into two valid
    syllables after first checking for illegal sequences.
    """
    for i in range(len(W) - 1):
        if W[i:i + 2] in ILLEGALS or W[i] in VOWELS and W[i + 1] in VOWELS:
            return 'ike'
    
    if is_valid_syllable(W):
        return 'pona'
    
    for i in range(len(W)):
        left, right = W[:i], W[i:]
        if is_valid_syllable(left) and is_valid_syllable(right):
            return 'pona'

    return 'ike'

def is_valid_syllable(s):
    if len(s) == 1:
        return s in VOWELS
    elif len(s) == 2:
        return s[0] in CONSONANTS and s[1] in VOWELS or \
               s[0] in VOWELS and s[1] == 'n'
    elif len(s) == 3:
        return s[0] in CONSONANTS and s[1] in VOWELS and s[2] == 'n'
    else:
        return False

def main():
    T = int(input())
    for _ in range(T):
        W = input()
        print(solve(W))

if __name__ == '__main__':
    main()

File: toki/solutions/toki_pair.cpp
#include <iostream>
#include <vector>

using namespace std;

string CONSONANTS = "mnptkswjl", VOWELS = "aeiou";
string ILLEGALS = "wu wo ji ti nn nm";

#define CONTAINS(text, pattern) text.find(pattern) != string::npos

bool is_valid_syllable(string s);

/**
 * Considers if W is a valid syllable or if W can be broken into two valid
 * syllables after first checking for illegal sequences.
 */
string solve(string W) {
    for (int i = 0; i < W.length() - 1; i++) {
        if (CONTAINS(ILLEGALS, W.substr(i, 2)) ||
            CONTAINS(VOWELS, W[i]) && CONTAINS(VOWELS, W[i + 1])) {
            return "ike";
        }
    }
    
    if (is_valid_syllable(W)) {
        return "pona";
    }
    
    for (int i = 0; i < W.length(); i++) {
        string left = W.substr(0, i), right = W.substr(i);
        if (is_valid_syllable(left) && is_valid_syllable(right)) {
            return "pona";
        }
    }
    
    return "ike";
}

bool is_valid_syllable(string s) {
    switch (s.length()) {
        case 1:
            return CONTAINS(VOWELS, s[0]);
        case 2:
            return CONTAINS(CONSONANTS, s[0]) && CONTAINS(VOWELS, s[1]) ||
                   CONTAINS(VOWELS, s[0]) && s[1] == 'n';
        case 3:
            return CONTAINS(CONSONANTS, s[0]) &&
                   CONTAINS(VOWELS, s[1]) &&
                   s[2] == 'n';
        default:
            return false;
    }
}

int main() {
    int T;
    cin >> T;
    for (int i = 0; i < T; i++) {
        string W;
        cin >> W;
        cout << solve(W) << '\n';
    }
}
